File: scripts/extract_all.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
递归解压脚本
支持递归解压所有压缩文件，并显示进度
"""

import os
import sys
import subprocess
import shutil
from pathlib import Path
from typing import List, Set

# 解压密码
PASSWORD = "cosergirl.com"

# 支持的压缩文件扩展名
COMPRESSED_EXTENSIONS = {'.7z', '.zip', '.rar', '.tar', '.gz', '.bz2', '.xz', '.tar.gz', '.tar.bz2', '.tar.xz'}


def is_compressed_file(file_path: Path) -> bool:
    """检查文件是否为压缩文件"""
    return file_path.suffix.lower() in COMPRESSED_EXTENSIONS or any(
        file_path.name.lower().endswith(ext) for ext in COMPRESSED_EXTENSIONS
    )


def extract_7z(file_path: Path, output_dir: Path, password: str = None) -> bool:
    """解压 7z 文件，优先使用系统 7z 命令，否则尝试 py7zr"""
    # 优先使用系统 7z 命令
    if shutil.which('7z') is not None:
        try:
            output_dir.mkdir(parents=True, exist_ok=True)
            cmd = ['7z', 'x', str(file_path), f'-o{output_dir}', '-y']
            if password:
                cmd.extend(['-p' + password])
            
            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                check=False
            )
            
            if result.returncode == 0:
                return True
            else:
                # 如果 7z 命令失败，继续尝试 py7zr
                pass
        except Exception as e:
            # 如果 7z 命令出错，继续尝试 py7zr
            pass
    
    # 尝试使用 py7zr 库
    try:
        try:
            import py7zr
        except ImportError:
            print(f"\n错误: 未安装 py7zr 库，且系统未找到 7z 命令")
            print("请选择以下方式之一安装:")
            print("  1. 安装 py7zr: pip3 install py7zr")
            print("  2. 安装 7z 工具: brew install p7zip  (需要 Homebrew)")
            return False
        
        # 确保输出目录存在
        output_dir.mkdir(parents=True, exist_ok=True)
        
        with py7zr.SevenZipFile(file_path, mode='r', password=password) as archive:
            archive.extractall(path=output_dir)
        
        return True
    except Exception as e:
        print(f"错误: 解压 {file_path.name} 时发生异常: {e}")
        return False


def extract_zip(file_path: Path, output_dir: Path, password: str = None) -> bool:
    """解压 ZIP 文件"""
    try:
        import zipfile
        with zipfile.ZipFile(file_path, 'r') as zip_ref:
            if password:
                zip_ref.setpassword(password.encode('utf-8'))
            zip_ref.extractall(output_dir)
        return True
    except Exception as e:
        print(f"错误: 解压 {file_path.name} 时发生异常: {e}")
        return False


def extract_file(file_path: Path, output_dir: Path, password: str = None) -> bool:
    """根据文件类型选择解压方法"""
    ext = file_path.suffix.lower()
    
    if ext == '.7z':
        return extract_7z(file_path, output_dir, password)
    elif ext == '.zip':
        return extract_zip(file_path, output_dir, password)
    else:
        # 对于其他格式，尝试使用 7z（它支持多种格式）
        return extract_7z(file_path, output_dir, password)


def find_compressed_files(directory: Path) -> List[Path]:
    """查找目录中的所有压缩文件"""
    compressed_files = []
    for file_path in directory.rglob('*'):
        if file_path.is_file() and is_compressed_file(file_path):
            compressed_files.append(file_path)
    return compressed_files


def recursive_extract(source_dir: Path, base_output_dir: Path = None, processed: Set[Path] = None):
    """
    递归解压所有压缩文件
    
    Args:
        source_dir: 源目录
        base_output_dir: 基础输出目录（如果为 None，则在源文件同级目录创建解压文件夹）
        processed: 已处理的文件集合（用于避免重复处理）
    """
    if processed is None:
        processed = set()
    
    if base_output_dir is None:
        base_output_dir = source_dir
    
    # 查找所有压缩文件
    compressed_files = find_compressed_files(source_dir)
    
    if not compressed_files:
        print("未找到压缩文件")
        return
    
    total_files = len(compressed_files)
    current = 0
    success_count = 0
    fail_count = 0
    
    print(f"\n找到 {total_files} 个压缩文件，开始解压...")
    print("=" * 60)
    
    while compressed_files:
        file_path = compressed_files.pop(0)
        
        # 跳过已处理的文件
        if file_path in processed:
            continue
        
        current += 1
        file_size = file_path.stat().st_size / (1024 * 1024)  # MB
        print(f"\n[{current}/{total_files}] 正在解压: {file_path.name} ({file_size:.2f} MB)")
        print(f"  路径: {file_path}")
        
        # 创建输出目录（在源文件同级目录）
        output_dir = file_path.parent / f"{file_path.stem}_extracted"
        
        # 解压文件
        success = extract_file(file_path, output_dir, PASSWORD)
        
        if success:
            print(f"  ✓ 解压成功 -> {output_dir}")
            processed.add(file_path)
            success_count += 1
            
            # 查找解压目录中的新压缩文件
            new_compressed = find_compressed_files(output_dir)
            for new_file in new_compressed:
                if new_file not in processed:
                    compressed_files.append(new_file)
                    total_files += 1
                    print(f"  → 发现新压缩文件: {new_file.name}")
        else:
            print(f"  ✗ 解压失败: {file_path.name}")
            processed.add(file_path)  # 标记为已处理，避免重复尝试
            fail_count += 1
        
        progress = current * 100 // total_files if total_files > 0 else 0
        print(f"  进度: {current}/{total_files} ({progress}%) | 成功: {success_count} | 失败: {fail_count}")
    
    print("\n" + "=" * 60)
    print(f"解压完成！")
    print(f"  总计: {current} 个文件")
    print(f"  成功: {success_count} 个")
    print(f"  失败: {fail_count} 个")


def main():
    """主函数"""
    # 检查是否有可用的解压工具
    has_7z = shutil.which('7z') is not None
    try:
        import py7zr
        has_py7zr = True
    except ImportError:
        has_py7zr = False
    
    if not has_7z and not has_py7zr:
        print("=" * 60)
        print("错误: 未找到解压工具")
        print("请选择以下方式之一安装:")
        print("  1. 安装 py7zr 库: pip3 install py7zr")
        print("  2. 安装 7z 工具: brew install p7zip  (需要 Homebrew)")
        print("=" * 60)
        sys.exit(1)
    
    if has_7z:
        print("使用系统 7z 工具进行解压")
    else:
        print("使用 py7zr 库进行解压")
    
    # 获取脚本所在目录的父目录（项目根目录）
    script_dir = Path(__file__).parent
    project_root = script_dir.parent
    
    # source 目录
    source_dir = project_root / 'source'
    
    if not source_dir.exists():
        print(f"错误: 源目录不存在: {source_dir}")
        sys.exit(1)
    
    print(f"源目录: {source_dir}")
    print(f"解压密码: {PASSWORD}")
    
    # 开始递归解压
    recursive_extract(source_dir)


if __name__ == '__main__':
    main()

File: scripts/extract_all.js
#!/usr/bin/env node
/**
 * 递归解压脚本 (Node.js 版本)
 * 支持递归解压所有压缩文件，并显示进度
 */

const fs = require("fs");
const path = require("path");
const { execSync, spawn } = require("child_process");
const { promisify } = require("util");

const readdir = promisify(fs.readdir);
const stat = promisify(fs.stat);
const mkdir = promisify(fs.mkdir);

// 解压密码
const PASSWORD = "cosergirl.com";

// 支持的压缩文件扩展名
const COMPRESSED_EXTENSIONS = [
  ".7z",
  ".zip",
  ".rar",
  ".tar",
  ".gz",
  ".bz2",
  ".xz",
  ".tar.gz",
  ".tar.bz2",
  ".tar.xz",
];

// 支持的图片文件扩展名
const IMAGE_EXTENSIONS = [
  ".jpg",
  ".jpeg",
  ".png",
  ".gif",
  ".bmp",
  ".webp",
  ".svg",
  ".ico",
  ".tiff",
  ".tif",
  ".heic",
  ".heif",
];

// 支持的视频文件扩展名
const VIDEO_EXTENSIONS = [
  ".mp4",
  ".avi",
  ".mov",
  ".mkv",
  ".wmv",
  ".flv",
  ".webm",
  ".m4v",
  ".mpg",
  ".mpeg",
  ".3gp",
  ".ts",
  ".mts",
];

/**
 * 检查文件是否为压缩文件（包括分卷压缩）
 */
function isCompressedFile(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  const name = path.basename(filePath).toLowerCase();

  // 检查标准压缩格式
  if (COMPRESSED_EXTENSIONS.some((ext) => name.endsWith(ext))) {
    return true;
  }

  // 检查分卷压缩文件（.001, .002, .003 等）
  // 只处理第一个分卷（.001），7z 会自动合并其他分卷
  const splitExtMatch = name.match(/\.(\d{3})$/);
  if (splitExtMatch) {
    const partNum = parseInt(splitExtMatch[1], 10);
    // 只处理 .001 文件（第一个分卷）
    if (partNum === 1) {
      // 检查是否是压缩格式的分卷（.7z.001, .rar.001 等）
      const baseName = name.substring(0, name.length - 4); // 移除 .001
      return COMPRESSED_EXTENSIONS.some((ext) => baseName.endsWith(ext));
    }
  }

  return false;
}

/**
 * 检查命令是否存在
 */
function which(command) {
  try {
    const result = execSync(`which ${command}`, {
      encoding: "utf8",
      stdio: ["ignore", "pipe", "ignore"],
    });
    return result.trim().length > 0;
  } catch {
    return false;
  }
}

/**
 * 使用 7z 命令解压文件
 */
function extract7z(filePath, outputDir, password = null) {
  return new Promise((resolve, reject) => {
    // 确保输出目录存在
    if (!fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir, { recursive: true });
    }

    const args = ["x", filePath, `-o${outputDir}`, "-y"];
    if (password) {
      args.push(`-p${password}`);
    }

    const child = spawn("7z", args, {
      stdio: ["ignore", "pipe", "pipe"],
    });

    let stdout = "";
    let stderr = "";

    child.stdout.on("data", (data) => {
      stdout += data.toString();
    });

    child.stderr.on("data", (data) => {
      stderr += data.toString();
    });

    child.on("close", (code) => {
      if (code === 0) {
        resolve(true);
      } else {
        reject(new Error(`7z 命令失败: ${stderr || stdout}`));
      }
    });

    child.on("error", (err) => {
      reject(err);
    });
  });
}

/**
 * 使用 unzip 命令解压 ZIP 文件
 */
function extractZip(filePath, outputDir, password = null) {
  return new Promise((resolve, reject) => {
    if (!fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir, { recursive: true });
    }

    const args = ["-o", filePath, "-d", outputDir];
    if (password) {
      args.push(`-P${password}`);
    }

    const child = spawn("unzip", args, {
      stdio: ["ignore", "pipe", "pipe"],
    });

    let stdout = "";
    let stderr = "";

    child.stdout.on("data", (data) => {
      stdout += data.toString();
    });

    child.stderr.on("data", (data) => {
      stderr += data.toString();
    });

    child.on("close", (code) => {
      if (code === 0) {
        resolve(true);
      } else {
        // unzip 可能返回非 0 但实际成功，检查输出
        if (stdout.includes("extracting") || stdout.includes("inflating")) {
          resolve(true);
        } else {
          reject(new Error(`unzip 命令失败: ${stderr || stdout}`));
        }
      }
    });

    child.on("error", (err) => {
      reject(err);
    });
  });
}

/**
 * 获取文件的基础扩展名（处理分卷压缩）
 */
function getBaseExtension(filePath) {
  const name = path.basename(filePath).toLowerCase();

  // 检查是否是分卷文件（.001, .002 等）
  const splitExtMatch = name.match(/\.(\d{3})$/);
  if (splitExtMatch) {
    // 移除分卷后缀，获取基础扩展名
    const baseName = name.substring(0, name.length - 4);
    const baseExt = path.extname(baseName).toLowerCase();
    return baseExt;
  }

  return path.extname(filePath).toLowerCase();
}

/**
 * 根据文件类型选择解压方法
 */
async function extractFile(filePath, outputDir, password = null) {
  const ext = getBaseExtension(filePath);
  const has7z = which("7z");
  const hasUnzip = which("unzip");

  try {
    // ZIP 文件：优先使用 unzip，否则使用 7z
    if (ext === ".zip") {
      if (hasUnzip) {
        return await extractZip(filePath, outputDir, password);
      } else if (has7z) {
        return await extract7z(filePath, outputDir, password);
      } else {
        throw new Error("未找到可用的解压工具 (需要 7z 或 unzip)");
      }
    }
    // 其他格式（.7z, .rar 等，包括分卷）：必须使用 7z
    else {
      if (has7z) {
        return await extract7z(filePath, outputDir, password);
      } else {
        throw new Error(
          `解压 ${ext} 格式需要 7z 工具。请运行: brew install p7zip`
        );
      }
    }
  } catch (error) {
    throw error;
  }
}

/**
 * 检查文件是否为图片或视频
 */
function isMediaFile(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  return IMAGE_EXTENSIONS.includes(ext) || VIDEO_EXTENSIONS.includes(ext);
}

/**
 * 检查文件夹是否只包含图片或视频文件（递归检查子文件夹）
 */
async function containsOnlyMediaFiles(dirPath) {
  try {
    const entries = await readdir(dirPath);

    for (const entry of entries) {
      const fullPath = path.join(dirPath, entry);
      const stats = await stat(fullPath);

      if (stats.isDirectory()) {
        // 递归检查子文件夹
        const subDirOnlyMedia = await containsOnlyMediaFiles(fullPath);
        if (!subDirOnlyMedia) {
          return false;
        }
      } else if (stats.isFile()) {
        // 检查文件是否为媒体文件
        if (!isMediaFile(fullPath)) {
          // 如果不是媒体文件，检查是否是压缩文件（需要继续解压）
          if (!isCompressedFile(fullPath)) {
            // 既不是媒体文件也不是压缩文件，返回 false
            return false;
          }
          // 如果是压缩文件，也返回 false（需要继续解压）
          return false;
        }
      }
    }

    return true;
  } catch (error) {
    console.error(`警告: 无法检查目录 ${dirPath}: ${error.message}`);
    return false;
  }
}

/**
 * 查找第一层包含媒体文件的文件夹
 * 返回这些文件夹的路径数组
 */
async function findFirstLevelMediaFolders(dirPath) {
  const mediaFolders = [];

  try {
    const entries = await readdir(dirPath);

    for (const entry of entries) {
      const fullPath = path.join(dirPath, entry);
      const stats = await stat(fullPath);

      if (stats.isDirectory()) {
        // 检查这个直接子文件夹是否只包含媒体文件
        const onlyMedia = await containsOnlyMediaFiles(fullPath);
        if (onlyMedia) {
          mediaFolders.push(fullPath);
        }
      }
    }
  } catch (error) {
    console.error(`警告: 无法扫描目录 ${dirPath}: ${error.message}`);
  }

  return mediaFolders;
}

/**
 * 递归查找目录中的所有压缩文件
 */
async function findCompressedFiles(directory) {
  const compressedFiles = [];

  async function scanDir(dir) {
    try {
      const entries = await readdir(dir);

      for (const entry of entries) {
        const fullPath = path.join(dir, entry);
        const stats = await stat(fullPath);

        if (stats.isDirectory()) {
          await scanDir(fullPath);
        } else if (stats.isFile() && isCompressedFile(fullPath)) {
          compressedFiles.push(fullPath);
        }
      }
    } catch (error) {
      // 忽略权限错误等
      console.error(`警告: 无法扫描目录 ${dir}: ${error.message}`);
    }
  }

  await scanDir(directory);
  return compressedFiles;
}

/**
 * 格式化文件大小
 */
function formatFileSize(bytes) {
  const mb = bytes / (1024 * 1024);
  return mb.toFixed(2);
}

/**
 * 递归解压所有压缩文件
 */
async function recursiveExtract(sourceDir, uploadDir) {
  const processed = new Set();
  let compressedFiles = await findCompressedFiles(sourceDir);

  if (compressedFiles.length === 0) {
    console.log("未找到压缩文件");
    return;
  }

  let totalFiles = compressedFiles.length;
  let current = 0;
  let successCount = 0;
  let failCount = 0;

  console.log(`\n找到 ${totalFiles} 个压缩文件，开始解压...`);
  console.log("=".repeat(60));

  while (compressedFiles.length > 0) {
    const filePath = compressedFiles.shift();

    // 跳过已处理的文件
    if (processed.has(filePath)) {
      continue;
    }

    current++;
    const stats = fs.statSync(filePath);
    const fileSize = formatFileSize(stats.size);
    const fileName = path.basename(filePath);

    console.log(
      `\n[${current}/${totalFiles}] 正在解压: ${fileName} (${fileSize} MB)`
    );
    console.log(`  路径: ${filePath}`);

    // 计算输出目录：先解压到临时位置（源文件同级目录）
    const dir = path.dirname(filePath);
    const baseExt = getBaseExtension(filePath);
    let fileNameWithoutExt = path.basename(filePath, baseExt);
    if (fileNameWithoutExt.match(/\.\d{3}$/)) {
      fileNameWithoutExt = fileNameWithoutExt.replace(/\.\d{3}$/, "");
    }

    // 临时解压目录（在源文件同级）
    const tempOutputDir = path.join(dir, `${fileNameWithoutExt}_extracted`);

    try {
      await extractFile(filePath, tempOutputDir, PASSWORD);
      console.log(`  ✓ 解压成功 -> ${tempOutputDir}`);
      processed.add(filePath);
      successCount++;

      // 查找第一层包含媒体文件的文件夹
      const mediaFolders = await findFirstLevelMediaFolders(tempOutputDir);

      // 将找到的媒体文件夹直接移动到 upload 目录
      for (const mediaFolder of mediaFolders) {
        const folderName = path.basename(mediaFolder);
        const finalOutputDir = path.join(uploadDir, folderName);

        // 检查目标文件夹是否已存在，如果存在则添加序号
        let targetDir = finalOutputDir;
        let counter = 1;
        while (fs.existsSync(targetDir)) {
          targetDir = path.join(uploadDir, `${folderName}_${counter}`);
          counter++;
        }

        // 移动文件夹到 upload
        fs.renameSync(mediaFolder, targetDir);
        console.log(`  → 移动到 upload: ${targetDir}`);
      }

      // 检查是否还有压缩文件需要处理
      const newCompressed = await findCompressedFiles(tempOutputDir);
      if (newCompressed.length > 0) {
        console.log(`  → 发现 ${newCompressed.length} 个压缩文件，继续处理...`);
        for (const newFile of newCompressed) {
          if (!processed.has(newFile)) {
            compressedFiles.push(newFile);
            totalFiles++;
            console.log(`  → 发现新压缩文件: ${path.basename(newFile)}`);
          }
        }
      } else if (mediaFolders.length === 0) {
        // 如果没有找到第一层媒体文件夹，且没有压缩文件，检查是否整个文件夹只包含媒体文件（直接文件，不是子文件夹）
        const entries = await readdir(tempOutputDir);
        let hasSubDirs = false;

        for (const entry of entries) {
          const fullPath = path.join(tempOutputDir, entry);
          const stats = await stat(fullPath);
          if (stats.isDirectory()) {
            hasSubDirs = true;
            break;
          }
        }

        // 如果没有子文件夹，且只包含媒体文件，才移动整个文件夹
        if (!hasSubDirs) {
          const onlyMedia = await containsOnlyMediaFiles(tempOutputDir);
          if (onlyMedia) {
            // 整个文件夹只包含媒体文件，直接移动到 upload
            const folderName = path.basename(tempOutputDir);
            let targetDir = path.join(uploadDir, folderName);
            let counter = 1;
            while (fs.existsSync(targetDir)) {
              targetDir = path.join(uploadDir, `${folderName}_${counter}`);
              counter++;
            }
            fs.renameSync(tempOutputDir, targetDir);
            console.log(`  → 移动到 upload: ${targetDir}`);
          }
        }
      }
    } catch (error) {
      console.error(`  ✗ 解压失败: ${fileName}`);
      console.error(`    错误: ${error.message}`);
      processed.add(filePath); // 标记为已处理，避免重复尝试
      failCount++;
    }

    const progress =
      totalFiles > 0 ? Math.floor((current * 100) / totalFiles) : 0;
    console.log(
      `  进度: ${current}/${totalFiles} (${progress}%) | 成功: ${successCount} | 失败: ${failCount}`
    );
  }

  console.log("\n" + "=".repeat(60));
  console.log("解压完成！");
  console.log(`  总计: ${current} 个文件`);
  console.log(`  成功: ${successCount} 个`);
  console.log(`  失败: ${failCount} 个`);
}

/**
 * 主函数
 */
async function main() {
  // 检查是否有可用的解压工具
  const has7z = which("7z");
  const hasUnzip = which("unzip");

  if (!has7z && !hasUnzip) {
    console.log("=".repeat(60));
    console.log("错误: 未找到解压工具");
    console.log("请选择以下方式之一安装:");
    console.log("  1. 安装 7z 工具: brew install p7zip  (推荐)");
    console.log("  2. 安装 unzip: brew install unzip");
    console.log("=".repeat(60));
    process.exit(1);
  }

  if (has7z) {
    console.log("使用系统 7z 工具进行解压");
  } else {
    console.log("使用系统 unzip 工具进行解压");
  }

  // 获取脚本所在目录的父目录（项目根目录）
  const scriptDir = __dirname;
  const projectRoot = path.dirname(scriptDir);

  // source 目录
  const sourceDir = path.join(projectRoot, "source");

  if (!fs.existsSync(sourceDir)) {
    console.error(`错误: 源目录不存在: ${sourceDir}`);
    process.exit(1);
  }

  // upload 目录（使用当天日期命名）
  const now = new Date();
  const year = now.getFullYear();
  const month = String(now.getMonth() + 1).padStart(2, "0");
  const day = String(now.getDate()).padStart(2, "0");
  const dateStr = `${year}${month}${day}`;
  const uploadDir = path.join(projectRoot, dateStr);

  // 确保 upload 目录存在
  if (!fs.existsSync(uploadDir)) {
    fs.mkdirSync(uploadDir, { recursive: true });
    console.log(`创建输出目录: ${uploadDir}`);
  }

  console.log(`源目录: ${sourceDir}`);
  console.log(`输出目录: ${uploadDir}`);
  console.log(`解压密码: ${PASSWORD}`);

  // 开始递归解压
  try {
    await recursiveExtract(sourceDir, uploadDir);
  } catch (error) {
    console.error("发生错误:", error.message);
    process.exit(1);
  }
}

// 运行主函数
if (require.main === module) {
  main().catch((error) => {
    console.error("未捕获的错误:", error);
    process.exit(1);
  });
}

module.exports = { recursiveExtract, extractFile, findCompressedFiles };
